### PR TITLE
Simplify BlocPresentationMixin testing

### PR DESCRIPTION
**Pull request by KrzysztofMamak**

### CODE REVIEW

**Review by Albert221 — COMMENTED** [review]
Good job!

* **Approach 1.** Using stream makes it flexible. A developer can use `Stream.fromIterable` if they just need to emit some presentation event(s) straight away, or create a `StreamController` and pass the `streamController.stream` if they want to have control over when the event is emitted.
* **Approach 2.** This solution is the subset of the 1st approach, making it worse 
* **Approach 3.** This solution also forces the developer to use a `StreamController` even if they would like to simply emit something straight away. Also, managing the lifecycle (closing) of something that we do not own I'd consider an anti-pattern.
* **Approach 4.** This is interesting approach. I'm neither pro nor against this one. Would it be possible though to extend `MockBloc` instead of just `Mock`? We could then save ourselves from mocking the `stream`.

I've thought of a mix of Approach 1. and 2. AFAIK, if we don't even want to test the presentation events but have their listener somewhere in the widget tree then we need to mock the `presentation` anyway. Maybe we could have a `whenListenPresentation` that accepts only one positioned parameter, the bloc/cubit itself, and additionally it could accept two optinoal named parameters: `List<BlocPresentationEvent> events` and `StreamController<BlocPresentationEvent> `eventsController`. If a developer wants to fire event straight away they would use the former, if they need more control they'd use the latter. Maybe in rare cases both.

File: pubspec.yaml [hunk 1]
```diff
@@ -10,15 +10,16 @@ environment:
   flutter: '>=3.3.0'
 
 dependencies:
+  bloc_test: ^9.1.1
   flutter:
     sdk: flutter
   flutter_bloc: ^8.0.0
   flutter_hooks: ^0.18.1
+  mocktail: ^0.3.0
```
Comment (shilangyu) [inline pubspec.yaml:18]: Not sure how to feel about bringing test dependencies to normal dependencies

File: lib/src/test_utils/test_utils.dart [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export 'package:bloc_presentation/src/test_utils/mock_presentation_bloc.dart';
```
Comment (Albert221) [inline lib/src/test_utils/test_utils.dart:1]: This file is a public library I suppose, it should be in `lib/test_utils.dart`.

We can later decide whether we want to have it as a part of `bloc_presentation` or make it a separate `_test` package

File: lib/bloc_presentation.dart [hunk 3]
```diff
@@ -1,4 +1,5 @@
 export 'src/bloc_presentation_event.dart';
 export 'src/bloc_presentation_listener.dart';
 export 'src/bloc_presentation_mixin.dart';
+export 'src/test_utils/test_utils.dart';
```
Comment (Albert221) [inline lib/bloc_presentation.dart:4]: Ah, you're exporting the test_utils from here. That's fine

File: lib/src/test_utils/mock_presentation_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+  /// Adds [event] to presentation stream
+  void emitMockPresentationEvent(BlocPresentationEvent event) {
+    _presentationController.add(event);
+  }
+}
+
+/// Another variant of 2. approach.
+///
+/// This variant allows us to get rid of bloc_test dependency and unnecessary
```
Comment (Albert221) [inline lib/src/test_utils/mock_presentation_bloc.dart:49]: I think I don't have a problem with depending on `bloc_test`, as if a developer is testing the presentation, they more certainly are also testing the blocs

File: pubspec.yaml (hunk 1 above)
Comment (Albert221) [reply pubspec.yaml:18]: Would you rather have a separate `bloc_presentation_test` package?

File: lib/src/test_utils/when_listen_presentation.dart [hunk 5]
```diff
@@ -4,23 +4,20 @@ import 'package:bloc_presentation/src/bloc_presentation_event.dart';
...
-  StreamController<BlocPresentationEvent>? controller,
+  List<BlocPresentationEvent>? initialEvents,
 }) {
-  final effectiveStream =
-      Stream<BlocPresentationEvent>.fromIterable(events ?? []);
-
-  if (controller != null) {
-    controller.addStream(effectiveStream);
-  }
+  final presentationController = StreamController<BlocPresentationEvent>()
```
Comment (Albert221) [inline lib/src/test_utils/when_listen_presentation.dart:16]: Don't forget to close the controller in `tearDown` here

Comment (Albert221) [reply lib/src/test_utils/when_listen_presentation.dart:16]: `tearDown` can't be called from inside a test, where `whenListenPresentation` is usually called

**Review by Albert221 — APPROVED** [review]

File: pubspec.yaml (hunk 1 above)
Comment (KrzysztofMamak) [reply pubspec.yaml]: Finally, we decided to create a separate package `bloc_presentation_test` which will be included in this repo.

File: lib/src/test_utils/mock_presentation_bloc.dart (hunk 4 above)
Comment (KrzysztofMamak) [reply lib/src/test_utils/mock_presentation_bloc.dart]: bloc_test dependency is no longer needed due to the latest changes.

**Pull request merged**